datafusion/src/datasource/object_store/mod.rs

```diff
@@ -219,25 +220,42 @@ impl ObjectStoreRegistry {
 
     /// Get a suitable store for the URI based on it's scheme. For example:
     /// - URI with scheme `file://` or no schema will return the default LocalFS store
-    /// - URI with scheme `s3://` will return the S3 store if it's registered
-    /// Returns a tuple with the store and the path of the file in that store
-    /// (URI=scheme://path).
+    /// - URI with scheme `s3://host:port` will return the S3 store if it's registered
+    /// Returns a tuple with the store and the self-described uri of the file in that store
     pub fn get_by_uri<'a>(
         &self,
         uri: &'a str,
     ) -> Result<(Arc<dyn ObjectStore>, &'a str)> {
-        if let Some((scheme, path)) = uri.split_once("://") {
-            let stores = self.object_stores.read();
-            let store = stores
-                .get(&*scheme.to_lowercase())
-                .map(Clone::clone)
-                .ok_or_else(|| {
-                    DataFusionError::Internal(format!(
-                        "No suitable object store found for {}",
-                        scheme
-                    ))
-                })?;
-            Ok((store, path))
+        // We do not support the remote object store on Windows OS
```

Comment:
@alamb, I totally agree with you, since the uri components may differ for different object store instance. That's why I prefer to keep the whole uri info in the return value. However, here this PR mainly focus on the key of hash map for object stores. The scheme info is not enough. In general, a remote object store can be identified by scheme://host:port

Reply:
I think the distinction I am trying to draw is that the current Object Store API is mapped by scheme and it would be up to the object store implementation to figure out how to handle host/port information

So rather than having one `HDFSObjectStore` instance for `server1:8000` and a second `HDFSObjectStore` instance for `server2:8290`, there would be a single `HDFSObjectStore` that would need to know how to dispatch appropriately to the different server hosts / ports

The same basic pattern holds for file systems (for example, there is a single `LocalFileSyetem` instance even though the local file system might have different disks mounted to `/data` and `/data2`).

I think also it would hold for S3 and other types of object stores (where depending on the region you need to request to a different endpoint)

Reply:
I should point out that my reason for disliking adding `host` and `port` to the object store key is that it doesn't make sense for many types of object stores (such as `LocalFileSystem` or `S3` which have no notion of host/port). It seems like this change is too HDFS specific and can be accomplished in a different way

Reply:
Thanks @alamb. Finally I get your point. To use the object store, there are three hierarchies: ObjectStoreRegistry -> ObjectStore -> ObjectStoreInstance. The ObjectStore is just for managing one kind of object store rather than being the instance. Previously I misunderstood it as a real instance.

One more question, for some object store, they support many kinds of schemes. For example, HDFS support:
- file://
- hdfs://
- viewfs://

I still think the returned path should include the original scheme info rather than throw it away. Then for the object store instance, it will be able to deal with the path for different kinds of schemes. What do you think?

Reply:
> I still think the returned path should include the original scheme info rather than throw it away. Then for the object store instance, it will be able to deal with the path for different kinds of schemes. What do you think?

I think returning the entire path, rather than stripping away the `scheme` makes a lot of sense 👍